src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs

```diff
@@ -218,7 +218,117 @@ private static unsafe int IndexOfVectorized(ref byte searchSpace, byte value, in
         exitFixed:;
                 return offset;
             }
+        }
+
+        public static bool TryIndicesOf(this Span<byte> buffer, byte value, Span<int> indices, out int numberOfIndices)
+        {
+            var length = buffer.Length;
+            if (length == 0 || indices.Length == 0)
+            {
+                numberOfIndices = 0;
+                return false;
+            }
+
+            return TryIndicesOf(ref buffer.DangerousGetPinnableReference(), value, length, indices, out numberOfIndices);
+        }
+
+        public static bool TryIndicesOf(this ReadOnlySpan<byte> buffer, byte value, Span<int> indices, out int numberOfIndices)
+        {
+            var length = buffer.Length;
+            if (length == 0 || indices.Length == 0)
+            {
+                numberOfIndices = 0;
+                return false;
+            }
+
+            return TryIndicesOf(ref buffer.DangerousGetPinnableReference(), value, length, indices, out numberOfIndices);
+        }
+
+        private unsafe static bool TryIndicesOf(ref byte searchSpace, byte value, int length, Span<int> indices, out int numberOfIndices)
+        {
+            numberOfIndices = 0;
+
+            fixed (byte* pSearchSpace = &searchSpace)
+            {
+                var searchStart = pSearchSpace;
+                var offset = 0;
+
+                while (true)
+                {
+                    if (Vector.IsHardwareAccelerated)
+                    {
+                        // Check Vector lengths
+                        if (length - Vector<byte>.Count >= offset)
+                        {
+                            Vector<byte> values = GetVector(value);
+                            do
+                            {
+                                var vFlaggedMatches = Vector.Equals(Unsafe.Read<Vector<byte>>(searchStart + offset), values);
+                                if (!vFlaggedMatches.Equals(Vector<byte>.Zero))
+                                {
+                                    // Found match, reuse Vector values to keep register pressure low
+                                    values = vFlaggedMatches;
+                                    break;
+                                }
+
+                                offset += Vector<byte>.Count;
+                            } while (length - Vector<byte>.Count >= offset);
+
+                            // Found match? Perform secondary search outside out of loop, so above loop body is small
+                            if (length - Vector<byte>.Count >= offset)
+                            {
+                                // Find offset of first match
+                                offset += LocateFirstFoundByte(values);
+                                // goto rather than inline return to keep function smaller
+                                goto exitFixed;
```

Comment:
I think it's ok for the initial merge, but I think it would be worth playing with an implementation that does not restart the search after finding a match. And that is for two reasons:
1. Restarting work costs. I think this is partially the reason for bytes apart having such large impact on the perf.
2. Today, it does not pay off to try to 32-byte align the vectorized data. I think the reason is that we laways generate unaligned load instructions for SIMD registers. I have heard this will change at some point, and then we will have to change the implementation to 32-byte align and then restarting (which messes up alignment) would not be an option.